# Why: install one stable copy of the runtime SPI before Dart isolates start.
diff --git a/engine/src/flutter/lib/ui/BUILD.gn b/engine/src/flutter/lib/ui/BUILD.gn
index 33d3ff7179f..be0852d81b5 100644
--- a/engine/src/flutter/lib/ui/BUILD.gn
+++ b/engine/src/flutter/lib/ui/BUILD.gn
@@ -256,6 +256,8 @@ source_set("ui") {
     "painting/vertices.h",
     "plugins/callback_cache.cc",
     "plugins/callback_cache.h",
+    "spi_api.cc",
+    "spi_api.h",
     "semantics/semantics_flags.cc",
     "semantics/semantics_flags.h",
     "semantics/semantics_update.cc",
@@ -401,6 +403,7 @@ if (enable_unittests) {
       "painting/canvas_spi_unittests.cc",
       "painting/path_unittests.cc",
       "painting/single_frame_codec_unittests.cc",
+      "spi_api_unittests.cc",
       "semantics/semantics_update_builder_unittests.cc",
       "window/platform_configuration_unittests.cc",
       "window/platform_message_response_dart_port_unittests.cc",
diff --git a/engine/src/flutter/lib/ui/spi_api.cc b/engine/src/flutter/lib/ui/spi_api.cc
new file mode 100644
index 00000000000..58a83156cdd
--- /dev/null
+++ b/engine/src/flutter/lib/ui/spi_api.cc
@@ -0,0 +1,40 @@
+// Copyright 2026 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "flutter/lib/ui/spi_api.h"
+
+#include <algorithm>
+#include <cstring>
+#include <mutex>
+
+#include "flutter/fml/logging.h"
+
+namespace flutter {
+namespace {
+
+FlutterSpiEngineApi g_spi_api{};
+std::once_flag g_spi_api_install_flag;
+bool g_spi_api_installed = false;
+
+}  // namespace
+
+void InstallSpiApi(const FlutterSpiEngineApi& api) {
+  std::call_once(g_spi_api_install_flag, [&api]() {
+    const size_t struct_size =
+        std::min(api.struct_size, sizeof(FlutterSpiEngineApi));
+    std::memcpy(&g_spi_api, &api, struct_size);
+    g_spi_api.struct_size = struct_size;
+    g_spi_api_installed = true;
+  });
+
+  FML_DCHECK(g_spi_api.version_major == api.version_major);
+  FML_DCHECK(g_spi_api.version_minor == api.version_minor);
+}
+
+const FlutterSpiEngineApi& SpiApi() {
+  FML_DCHECK(g_spi_api_installed);
+  return g_spi_api;
+}
+
+}  // namespace flutter
diff --git a/engine/src/flutter/lib/ui/spi_api.h b/engine/src/flutter/lib/ui/spi_api.h
new file mode 100644
index 00000000000..82bb3830f64
--- /dev/null
+++ b/engine/src/flutter/lib/ui/spi_api.h
@@ -0,0 +1,21 @@
+// Copyright 2026 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef FLUTTER_LIB_UI_SPI_API_H_
+#define FLUTTER_LIB_UI_SPI_API_H_
+
+#include "flutter/runtime_spi/include/flutter_runtime_spi.h"
+
+namespace flutter {
+
+// Instalada pelo runtime Dart antes de qualquer isolate rodar; todas as chamadas devem passar a mesma ABI (major/minor).
+// Guarda uma cópia: as sub-tabelas do core são estáticas, a struct de cada engine não é.
+void InstallSpiApi(const FlutterSpiEngineApi& api);
+
+// FML_DCHECK de que InstallSpiApi rodou.
+const FlutterSpiEngineApi& SpiApi();
+
+}  // namespace flutter
+
+#endif  // FLUTTER_LIB_UI_SPI_API_H_
diff --git a/engine/src/flutter/lib/ui/spi_api_unittests.cc b/engine/src/flutter/lib/ui/spi_api_unittests.cc
new file mode 100644
index 00000000000..d9a5716119b
--- /dev/null
+++ b/engine/src/flutter/lib/ui/spi_api_unittests.cc
@@ -0,0 +1,34 @@
+// Copyright 2026 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "flutter/lib/ui/spi_api.h"
+#include "flutter/runtime_spi/spi_engine_api.h"
+
+#include "flutter/testing/testing.h"
+
+namespace flutter::testing {
+namespace {
+
+TEST(SpiApiTest, InstalledCopyOutlivesTheSource) {
+  {
+    const FlutterSpiEngineApi source_api = SpiEngineApi::Create();
+    InstallSpiApi(source_api);
+  }
+
+  EXPECT_EQ(SpiApi().scene, SpiEngineApi::Create().scene);
+  EXPECT_NE(SpiApi().struct_size, 0u);
+}
+
+TEST(SpiApiTest, SecondInstallKeepsTheFirst) {
+  InstallSpiApi(SpiEngineApi::Create());
+  const FlutterSpiSceneApi* first_scene = SpiApi().scene;
+  FlutterSpiEngineApi second_api = SpiEngineApi::Create();
+  second_api.scene = nullptr;
+  InstallSpiApi(second_api);
+
+  EXPECT_EQ(SpiApi().scene, first_scene);
+}
+
+}  // namespace
+}  // namespace flutter::testing
diff --git a/engine/src/flutter/runtime_spi/BUILD.gn b/engine/src/flutter/runtime_spi/BUILD.gn
index b2f1a7a230a..210cdcbbe27 100644
--- a/engine/src/flutter/runtime_spi/BUILD.gn
+++ b/engine/src/flutter/runtime_spi/BUILD.gn
@@ -28,6 +28,7 @@ source_set("spi_backend") {
     "spi_runtime_backend.h",
     "spi_runtime_backend_factory.h",
     "internal/spi_internal_bridges.h",
+    "spi_engine_api.h",
     "spi_scene.h",
   ]
   sources = [
diff --git a/engine/src/flutter/shell/common/dart_runtime/flutter_dart_runtime.cc b/engine/src/flutter/shell/common/dart_runtime/flutter_dart_runtime.cc
index 5ae511957ab..b82f1f02146 100644
--- a/engine/src/flutter/shell/common/dart_runtime/flutter_dart_runtime.cc
+++ b/engine/src/flutter/shell/common/dart_runtime/flutter_dart_runtime.cc
@@ -12,6 +12,7 @@
 
 #include "assets/native_assets.h"
 #include "flutter/fml/mapping.h"
+#include "flutter/lib/ui/spi_api.h"
 #include "flutter/lib/ui/window/key_data.h"
 #include "flutter/lib/ui/window/platform_message_response.h"
 #include "flutter/runtime/dart_vm.h"
@@ -353,6 +354,7 @@ DartRuntimeDelegateBridge& FlutterDartRuntime::delegate_bridge() {
 
 void FlutterDartRuntime::Start(const FlutterSpiEngineApi* engine_api,
                                FlutterSpiEngine* engine) {
+  InstallSpiApi(*engine_api);
   delegate_bridge_->SetEngineApi(engine_api, engine);
 }
 
